### PR TITLE
Additional VirtualService Hosts verify

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

fix: #15226

### CODE REVIEW

**Comment by howardjohn** [comment]
Not sure I understand, why is hosts an array if we can never have multiple? or am I missing something

**Comment by hzxuzhonghu** [comment]
>Not sure I understand, why is hosts an array if we can never have multiple? or am I missing something

My understanding is for internal k8s service, we can never have multiple host set. But for other mesh external service with multiple domains, we can.

File: pilot/pkg/model/validation.go [hunk 1]
```diff
@@ -1813,6 +1815,25 @@ func ValidateVirtualService(_, _ string, msg proto.Message) (errs error) {
 					errs = appendErrors(errs, fmt.Errorf("duplicate hosts in virtual service: %s & %s", hostI, hostJ))
 				}
 			}
+
+			if !strings.Contains(virtualService.Hosts[i], ".") {
+				svcShortNames = append(svcShortNames, virtualService.Hosts[i])
+			} else if strings.Contains(virtualService.Hosts[i], "."+namespace+"."+"svc") {
```
Comment (howardjohn) [inline pilot/pkg/model/validation.go]: are these valid:

a.foo.svc.cluster.local in namespace bar?

a.foo  (no svc.cluster.local suffix)?

Comment (hzxuzhonghu) [reply pilot/pkg/model/validation.go]: This is valid.

Comment (howardjohn) [reply pilot/pkg/model/validation.go]: So this check `virtualService.Hosts[i], "."+namespace+"."+"svc"` won't be sufficient then?

I don't understand what is valid and what isn't in VirtualService, and why we need to make this invalid. The current docs we have say nothing about not allowing multiple hosts.

I get the bug that was associated with this, but should we be rejecting these in validation or is there a bug in the config generation we need to fix?

Comment (hzxuzhonghu) [reply pilot/pkg/model/validation.go]: Can you look the test cases i added, it will help.

File: pilot/pkg/model/validation_test.go [hunk 2]
```diff
@@ -2606,6 +2606,30 @@ func TestValidateVirtualService(t *testing.T) {
...
+			}},
+		}, valid: false},
+		{name: "both service shotname and service FQDN specified", in: &networking.VirtualService{
+			Hosts: []string{"svc1", "svc2.default.svc.cluster.local"},
+			Http: []*networking.HTTPRoute{{
+				Route: []*networking.HTTPRouteDestination{{
+					Destination: &networking.Destination{Host: "foo.baz"},
+				}},
+			}},
+		}, valid: false},
```
Comment (howardjohn) [inline pilot/pkg/model/validation_test.go]: I don't understand. When is it valid to have multiple hosts? It seems like now you are saying it is never ok to have multiple hosts

Comment (hzxuzhonghu) [reply pilot/pkg/model/validation_test.go]: Sorry, it makes you confused, i will add multi hosts cases.

File: pilot/pkg/model/validation.go [hunk 3]
```diff
@@ -1813,6 +1815,25 @@ func ValidateVirtualService(_, _ string, msg proto.Message) (errs error) {
...
+				svcFQDNs = append(svcFQDNs, virtualService.Hosts[i])
+			}
+		}
+
+		// Check for multi services short name.
+		if len(svcShortNames) > 1 {
+			errs = appendErrors(errs, fmt.Errorf("multi service names specified in virtual service hosts: %v", virtualService.Hosts))
```
Comment (rshriram) [inline pilot/pkg/model/validation.go]: what does this check enforce?

File: pilot/pkg/model/validation_test.go [hunk 4]
```diff
@@ -2606,6 +2606,38 @@ func TestValidateVirtualService(t *testing.T) {
 				}},
 			}},
 		}, valid: false},
+		{name: "multi short names", in: &networking.VirtualService{
+			Hosts: []string{"svc1", "svc2"},
```
Comment (rshriram) [inline pilot/pkg/model/validation_test.go]: we dont allow shortnames at all do we? @frankbu ?

File: pilot/pkg/model/validation_test.go [hunk 5]
```diff
@@ -2606,6 +2606,38 @@ func TestValidateVirtualService(t *testing.T) {
...
+			Http: []*networking.HTTPRoute{{
+				Route: []*networking.HTTPRouteDestination{{
+					Destination: &networking.Destination{Host: "foo.baz"},
+				}},
+			}},
+		}, valid: false},
+		{name: "multi services FQDN", in: &networking.VirtualService{
+			Hosts: []string{"svc1.default.svc.cluster.local", "svc2.default.svc.cluster.local"},
```
Comment (rshriram) [inline pilot/pkg/model/validation_test.go]: why is this wrong?

**Comment by rshriram** [comment]
@hzxuzhonghu these fixes are not necessary [actually not valid] because the bug in question [ #15226 ] is not a bug, but by design. We do allow multiple services to be routed to a single destination if the user wishes to.

File: pilot/pkg/model/validation.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/validation.go]: Donot allow setting two diff k8s services, like svcA and svcB, which is meaningless.

File: pilot/pkg/model/validation_test.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/validation_test.go]: For k8s service, i think we should only allow one service name specified.  For mesh external services, this fields must be DNS domains?

**Comment by hzxuzhonghu** [comment]
>@hzxuzhonghu these fixes are not necessary [actually not valid] because the bug in question [ #15226 ] is not a bug, but by design. We do allow multiple services to be routed to a single destination if the user wishes to.

IC,  just a little confusing

**Pull request closed without merging**